core/src/main/scala/kafka/log/Log.scala

```
@@ -444,6 +457,23 @@ class Log(@volatile var dir: File,
     }
   }
 
+  /*
+   * Increment the log start offset if the provided offset is larger.
+   * And delete old segments whose largest offset < log start offset.
+   */
+  def maybeIncrementLogStartOffsetAndDeleteRecords(offset: Long) {
```

If we modify the log, we should probably coordinate this with the log cleaner. In LogManager.truncateTo(), we first abort the ongoing cleaning, then truncate and finally resume the cleaning. We probably should do the same here.

---

Good point. I will only delete non-compact topics in `maybeIncrementLogStartOffsetAndDeleteRecords` and let `LogCleaner` take care of compact topics.

---

So, DeleteRecordsRequest can't be applied on a compacted topic? Should we return an error code in the response in that case?

---

@junrao My apology, I missed your comment here. `DeleteRecordsrRequest` can actually be applied on a compacted topic. `DeleteRecordsrRequest` will increase the logStartOffset. When `LogCleaner` calls `log.deleteOldSegments()`, those log segments which breach the logStartOffset will be deleted.